File: impl/ts/src/cancel.ts
// Cooperative statement cancellation (spec/design/api.md §11.4).
//
// TS divergence (CLAUDE.md §2 — best experience per language, not a uniform-shape failure). Go and
// Rust have real threads, so another thread flips a token the cost meter polls MID-statement (the
// `guard()` checkpoint), interrupting a long-running statement at the next metering point. TypeScript
// runs on ONE event loop: nothing else — no timer, no I/O callback, no other thread — runs DURING a
// synchronous query()/execute(), so an AbortSignal's `aborted` state CANNOT change while a statement
// runs; it is frozen at the value it had on entry. Polling it at the meter checkpoint (the Go/Rust
// mechanism) would re-read that same value N times, so it is pointless here. TS therefore honors the
// signal at OPERATION BOUNDARIES only: the cancelable query/execute methods check it before any work
// and throw `57014 query_canceled` if it is already aborted — which usefully skips work for an
// already-canceled operation (e.g. a client that disconnected before the handler ran). Mid-statement
// cancellation would require the engine to become async (a streaming cursor that `await`s, §4); the
// boundary check is the forward-compatible seam for that day.
//
// The cancellation primitive is the platform `AbortSignal` (an `AbortController`), the same handle
// `fetch` and the web APIs use — so there is no custom token type (Rust ships `CancellationToken`
// only because it has no built-in equivalent). The meter and the SessionState are deliberately NOT
// wired, unlike Go/Rust: there is nothing for a synchronous run to poll.

import { engineError } from "./errors.ts";

// throwIfAborted throws `57014 query_canceled` when `signal` is already aborted, else returns. The
// boundary poll the cancelable host methods (Database/Session/Transaction `*Cancelable`) run before
// executing a statement. `undefined` (no signal supplied) is a no-op — zero overhead, the path every
// non-cancelable call and every conformance / cost test takes.
export function throwIfAborted(signal: AbortSignal | undefined): void {
  if (signal?.aborted) {
    throw engineError("query_canceled", "canceling statement due to user request");
  }
}

File: impl/rust/src/cancel.rs
//! Cooperative statement cancellation (spec/design/api.md §11.4).
//!
//! A [`CancellationToken`] is a clonable handle around a shared atomic flag. The host arms a clone
//! on the session that runs a statement (the cancelable query/execute methods on
//! [`Session`](crate::Session) / [`Database`](crate::Database) / [`Transaction`](crate::Transaction)),
//! and the statement's cost [`Meter`](crate::cost::Meter) consults it at every
//! [`guard`](crate::cost::Meter::guard) checkpoint — the same single chokepoint already run at the
//! unbounded-work points for `max_cost` (spec/design/cost.md §6). A flipped token therefore aborts a
//! **long-running** statement with `57014 query_canceled` at the next metering point, not only at the
//! cursor / entry boundary.
//!
//! Rust has real threads (CLAUDE.md §2), so the token is the whole mechanism: thread A runs the
//! query while thread B (a timeout, a request-cancel) flips a clone of the *same* token. No watcher
//! is needed — the token *is* the shared atomic the Go core spins up a goroutine to feed. It is the
//! Rust spelling of Go's `context.Context` cancellation; TS, which cannot preempt synchronous
//! execution, honors its `AbortSignal` only at boundaries (api.md §11.4).
//!
//! This rides the existing meter seam, so it touches no conformance / cost path: an un-armed meter
//! carries `None`, the `guard` short-circuits, and the cross-core cost determinism (CLAUDE.md §8) is
//! untouched. Cancellation is per-core unit-tested, never in the corpus — whether a cancel lands at
//! row 1k or 5k is timing (api.md §11.4, CLAUDE.md §10).

use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};

use crate::error::{EngineError, Result};
use crate::sqlstate::SqlState;

/// A clonable cancellation handle (spec/design/api.md §11.4). Cloning shares the underlying flag
/// (`Arc<AtomicBool>`), so a clone handed to another thread cancels the statement the original is
/// arming. Cheap to clone and to poll (a single relaxed atomic load on the hot path).
#[derive(Clone, Default)]
pub struct CancellationToken(Arc<AtomicBool>);

impl CancellationToken {
    /// A fresh, un-cancelled token.
    pub fn new() -> Self {
        CancellationToken::default()
    }

    /// Request cancellation. Idempotent, and callable from any thread — the statement arming a clone
    /// of this token aborts `57014` at its next [`Meter::guard`](crate::cost::Meter::guard).
    pub fn cancel(&self) {
        // Relaxed is sufficient: the flag synchronizes no other memory, only signals "abort at the
        // next checkpoint." The abort still surfaces as a clean error on the running thread.
        self.0.store(true, Ordering::Relaxed);
    }

    /// Whether cancellation has been requested.
    pub fn is_cancelled(&self) -> bool {
        self.0.load(Ordering::Relaxed)
    }

    /// The cheap boundary poll (the analog of Go's `ctxErr`): `Err(57014)` if already cancelled, else
    /// `Ok`. Used at the API entry of a cancelable call before any work, complementing the in-statement
    /// meter `guard`.
    pub(crate) fn check(&self) -> Result<()> {
        if self.is_cancelled() {
            Err(EngineError::new(
                SqlState::QueryCanceled,
                "canceling statement due to user request",
            ))
        } else {
            Ok(())
        }
    }
}
